### PR TITLE
Switch extension controller to `logr` and streamline/cleanup logs

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity logging
/kind enhancement

**What this PR does / why we need it**:
This PR 

- streamlines the log level when an object is no longer found
- streamlines the finalizer handling (logs + errors)
- uses `logf.FromContext(ctx)` in all controllers in `./extensions/...`
- drops usage of global `logf.Log` in favor of local reconciler loggers

**Which issue(s) this PR fixes**:
Part of #4251 

**Release note**:

```breaking dependency
All `Actuator` interfaces for extension controllers have been extended and now receive a `logr.Logger` passed from the reconciler with the proper context of the reconciled object.
```

### CODE REVIEW

File: extensions/pkg/controller/csimigration/reconciler.go [hunk 1]
```diff
@@ -66,37 +64,41 @@ func (r *reconciler) InjectClient(client client.Client) error {
...
 	}
 
 	if cluster.DeletionTimestamp != nil {
 		return reconcile.Result{}, nil
 	}
 
-	r.logger.Info("CSI migration controller got called with cluster", "csimigration", kutil.ObjectName(cluster))
-
-	return r.reconcile(ctx, cluster, shoot)
+	log.Info("CSI migration controller got called with cluster")
```
Comment (plkokanov) [inline extensions/pkg/controller/csimigration/reconciler.go:94]: The log now does not include the cluster name and will instead print the shoot object key. Perhaps ot should be changed to:
 ```suggestion
 	log.Info("CSI migration controller got called for shoot")
 ```
 
 Additionally all subsequent logs will print the shoot key instead of the cluster, but I guess that is ok?

File: extensions/pkg/controller/extension/reconciler.go [hunk 2]
```diff
@@ -129,69 +128,85 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
 	}
 
 	return reconcile.Result{}, nil
 }
 
-func (r *reconciler) delete(ctx context.Context, ex *extensionsv1alpha1.Extension) (reconcile.Result, error) {
+func (r *reconciler) delete(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) (reconcile.Result, error) {
 	if !controllerutil.ContainsFinalizer(ex, r.finalizerName) {
-		r.logger.Info("Deleting extension causes a no-op as there is no finalizer", "extension", kutil.ObjectName(ex))
+		log.Info("Deleting extension causes a no-op as there is no finalizer")
```
Comment (plkokanov) [inline extensions/pkg/controller/extension/reconciler.go]:
```suggestion
		log.Info("Deleting Extension causes a no-op as there is no finalizer")
```

File: extensions/pkg/controller/network/reconciler.go [hunk 3]
```diff
@@ -116,112 +113,145 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
 
-	if err := r.statusUpdater.Processing(ctx, network, gardencorev1beta1.LastOperationTypeDelete, "Deleting the network"); err != nil {
+	if err := r.statusUpdater.Processing(ctx, network, gardencorev1beta1.LastOperationTypeDelete, "Deleting the Network"); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	r.logger.Info("Starting the deletion of network", "network", kutil.ObjectName(network))
-	if err := r.actuator.Delete(ctx, network, cluster); err != nil {
-		_ = r.statusUpdater.Error(ctx, network, reconcilerutils.ReconcileErrCauseOrErr(err), gardencorev1beta1.LastOperationTypeDelete, "Error deleting network")
+	log.Info("Starting the deletion of network")
```
Comment (plkokanov) [inline extensions/pkg/controller/network/reconciler.go]:
```suggestion
	log.Info("Starting the deletion of Network")
```

File: extensions/pkg/controller/worker/controller.go [hunk 4]
```diff
@@ -33,9 +32,9 @@ const (
 	// FinalizerName is the worker controller finalizer.
 	FinalizerName = "extensions.gardener.cloud/worker"
 	// ControllerName is the name of the controller.
-	ControllerName = "worker_controller"
+	ControllerName = "worker"
 	// StateUpdatingControllerName is the name of the controller responsible for updating the worker's state.
-	StateUpdatingControllerName = "worker_state_controller"
+	StateUpdatingControllerName = "state"
```
Comment (plkokanov) [inline extensions/pkg/controller/worker/controller.go]: I'm wondering if the name here shouldn't be something in the lines of `worker.state` or to at least result in `worker.state` when the logs for the state controller are printed. This is in case we have to add a state reconciler to some other controller in the future.

**Review by plkokanov — CHANGES_REQUESTED** [review]
Halfway through the commits.

File: extensions/pkg/controller/backupentry/reconciler.go [hunk 5]
```diff
@@ -149,8 +152,12 @@ func (r *reconciler) reconcile(
 	if err != nil {
 		return reconcile.Result{}, fmt.Errorf("failed to get backup entry secret: %+v", err)
 	}
-	if err := controllerutils.EnsureFinalizer(ctx, r.reader, r.client, secret, FinalizerName); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to ensure finalizer on backup entry secret: %+v", err)
+
+	if !controllerutil.ContainsFinalizer(secret, FinalizerName) {
+		log.Info("Adding finalizer to secret", "secret", client.ObjectKeyFromObject(secret))
+		if err := controllerutils.AddFinalizers(ctx, r.client, secret, FinalizerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
```
Comment (timebertt) [inline extensions/pkg/controller/backupentry/reconciler.go]: Sometimes you log `failed to add finalizer to secret` sometimes `failed to add finalizer`.
Shall we make it consistent?

File: pkg/controllerutils/finalizers.go [hunk 6]
```diff
@@ -19,87 +19,37 @@ import (
 
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	"k8s.io/client-go/util/retry"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// PatchAddFinalizers adds the given finalizers to the object via a merge patch request with optimistic locking.
-func PatchAddFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
+// AddFinalizers ensures that the given finalizer is present in the given object with exponential backoff and optimistic
```
Comment (timebertt) [inline pkg/controllerutils/finalizers.go]: `AddFinalizers` and `AddFinalizers` don't perform exponential backoff themselves. 
Also, they always issue a patch contradicting to the doc string.
Can you adapt the doc strings to avoid confusion?

File: pkg/controllerutils/finalizers.go [hunk 7]
```diff
@@ -19,87 +19,37 @@ import (
...
+func RemoveFinalizers(ctx context.Context, writer client.Writer, obj client.Object, finalizers ...string) error {
+	return client.IgnoreNotFound(patchFinalizers(ctx, writer, obj, mergeFromWithOptimisticLock, controllerutil.RemoveFinalizer, finalizers...))
+}
+
+// RemoveAllFinalizers ensures that the given object has no finalizers with exponential backoff.
+// If any finalizers are set, it removes them and issues a patch.
+func RemoveAllFinalizers(ctx context.Context, writer client.Writer, obj client.Object) error {
+	beforePatch := obj.DeepCopyObject().(client.Object)
+	obj.SetFinalizers(nil)
+	return client.IgnoreNotFound(writer.Patch(ctx, obj, mergeFromWithOptimisticLock(beforePatch)))
```
Comment (timebertt) [inline pkg/controllerutils/finalizers.go]: We could even drop optimistic locking here :)
```suggestion
	return client.IgnoreNotFound(writer.Patch(ctx, obj, mergeFrom(beforePatch)))
```

File: cmd/gardener-admission-controller/app/gardener_admission_controller.go [hunk 8]
```diff
@@ -179,23 +179,23 @@ func (o *options) run(ctx context.Context) error {
 		return err
 	}
 
-	namespaceValidationHandler, err := namespacedeletion.New(ctx, logf.Log.WithName(namespacedeletion.HandlerName), mgr.GetCache())
+	namespaceValidationHandler, err := namespacedeletion.New(ctx, log.WithName(namespacedeletion.HandlerName), mgr.GetCache())
```
Comment (timebertt) [inline cmd/gardener-admission-controller/app/gardener_admission_controller.go]: Instead of inheriting from the global logger, I propose to use a logger with name `webhook` and add the name via `WithValues("webhook", name)`.
This is similar to what controller-runtime does for controllers.

However, the better solution would be to contribute a small PR to controller-runtime that calls `logf.IntoContext` in the webhook server as well, so that one can simply use `logf.FromContext` in webhook handlers.

File: extensions/pkg/controller/csimigration/reconciler.go (hunk 1 above)
Comment (rfranzke) [reply extensions/pkg/controller/csimigration/reconciler.go:94]: It will print have both keys (`cluster` because that's the object the controller reconciles, and `shoot` because I added it manually). I think it's still fine like this?

**Review by timebertt — COMMENTED** [review]
Wow, this PR was a lot of effort! Nice work, highly appreciated ❤️ 
Special thanks for the harmonization around finalizer handling.

Can you add a breaking release note for extensions about the added log parameters in the library?

File: extensions/pkg/controller/csimigration/reconciler.go (hunk 1 above)
Comment (plkokanov) [reply extensions/pkg/controller/csimigration/reconciler.go:94]: Ah yes you're right. I forgot that the CSI migration controller reconciles `cluster` resources. It's fine as it currently is

**Review by timebertt — APPROVED** [review]
Sweet!
/lgtm

**Review by plkokanov — APPROVED** [review]
Amazing PR :)
/lgtm

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go [hunk 9]
```diff
@@ -211,8 +210,8 @@ func NewGardener(ctx context.Context, cfg *config.ControllerManagerConfiguration
 		return nil, fmt.Errorf("error instantiating zap logger: %w", err)
 	}
 
-	runtimelog.SetLogger(log)
-	klog.SetLogger(log)
+	// set the logger used by sigs.k8s.io/controller-runtime
+	logf.SetLogger(log)
```
Comment (timebertt) [inline cmd/gardener-controller-manager/app/gardener_controller_manager.go]: Why did you remove `klog.SetLogger(log)`?

File: cmd/gardenlet/app/gardenlet.go [hunk 10]
```diff
@@ -243,8 +242,8 @@ func NewGardenlet(ctx context.Context, cfg *config.GardenletConfiguration) (*Gar
 		return nil, fmt.Errorf("error instantiating zap logger: %w", err)
 	}
 
-	runtimelog.SetLogger(log)
-	klog.SetLogger(log)
+	// set the logger used by sigs.k8s.io/controller-runtime
+	logf.SetLogger(log)
```
Comment (timebertt) [inline cmd/gardenlet/app/gardenlet.go]: Same here

**Review by timebertt — APPROVED** [review]
/lgtm
